src/collection/shared/generic_collector.py

```diff
@@ -0,0 +1,174 @@
+"""
+Generic class for data collection from Solana chain
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import List, Tuple
+import logging
+import os
+import time
+
+import solana.rpc.api
+from solana.exceptions import SolanaRpcException
+from solders.signature import Signature
+from solders.transaction_status import EncodedTransactionWithStatusMeta, UiConfirmedBlock
+
+LOG = logging.getLogger(__name__)
+
+
+def log_performance_time(logger = LOG):
+    """
+    Decorator for logging performance time of subjected function.
+    """
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            start_time = time.perf_counter()
+            log_message = f"START: function '{func.__name__}'."
+            # Log the start of the function
+            logger.info(log_message)
+
+            # Call the actual function
+            result = func(*args, **kwargs)
+
+            # Calculate elapsed time and log the end of the function
+            elapsed_time = time.perf_counter() - start_time
+            logger.info(f"DONE: function '{func.__name__}' in {elapsed_time:.2f} seconds")
+
+            return result
+
+        return wrapper
+    return decorator
+
+
+@dataclass
+class SolanaTransaction:
+    block_time: int
+    block_number: int
+    tx_body: EncodedTransactionWithStatusMeta
+
+    def sources(self, ppks: List[str]) -> List[str]:
+        """
+        Identify transaction source by matching present public keys with provided protocols' public keys
+        """
+        relevant_sources = [
+            k for k in ppks
+            if k in [str(i.pubkey) for i in self.tx_body.transaction.message.account_keys]  # type: ignore
+        ]
+        if not relevant_sources:
+            LOG.error(f"Transaction `{self.tx_body.transaction.signatures[0]}`"
+                      f" does not contain any relevant public keys.")
+
+        return relevant_sources
+
+    @property
+    def signature(self) -> Signature:
+        """ Returns transaction signature. """
+        return self.tx_body.transaction.signatures[0]
+
+
+class GenericSolanaConnector(ABC):
+    """
+    Abstract class with methods, that every Solana connector should implement.
+    """
+
+    def __init__(self):
+        self._get_rpc_url()
+        self.solana_client: solana.rpc.api.Client = solana.rpc.api.Client(self.authenticated_rpc_url)
+
+        self.assignment: List[int] = list()
+
+        # attribute for storing transactions before assigning to db.
+        self.rel_transactions: List[SolanaTransaction] = list()
+
+        # rate limiter
+        self.rate_limit: int = int(os.getenv("RATE_LIMIT", ""))
```

Comment:
This could be consistent with `self._get_rpc_url`.

Reply:
agree. i think following would work:
```
    def _set_rate_limit(self) -> None:
        rate_limit = os.getenv("RATE_LIMIT", "")
        if not rate_limit:
            LOG.error("Rate limit was not found in environment variables and set to 1.")
            self.rate_limit = 1
            return
        self.rate_limit = int(rate_limit)
 ```
 also 1 is set as default.

Reply:
:+1: 

Could we use `None` instead of `""` as the default?

Reply:
done